File: python-client/src/hammerspoon_macapi/namespaces/apps.py
"""Typed application observation and control methods."""

from __future__ import annotations

from typing import TYPE_CHECKING, cast

from ..models import ApplicationInfo

if TYPE_CHECKING:
    from ..client import MacAPI


class AppsClient:
    """Namespace wrapper for the ``apps.*`` RPC methods."""

    def __init__(self, api: MacAPI) -> None:
        self._api = api

    async def list(self) -> list[ApplicationInfo]:
        """Return all currently running applications."""
        return await self._api.call("apps.list", result_type=list[ApplicationInfo])

    async def frontmost(self) -> ApplicationInfo | None:
        """Return the frontmost application, if one is available."""
        return cast(
            ApplicationInfo | None,
            await self._api.call("apps.frontmost", result_type=ApplicationInfo | None),
        )

    async def open(self, bundle_id: str) -> ApplicationInfo:
        """Launch or focus an application by bundle identifier."""
        return await self._api.call(
            "apps.open", {"bundle_id": bundle_id}, result_type=ApplicationInfo
        )

    async def focus(self, bundle_id: str) -> None:
        """Activate a running application."""
        await self._api.call("apps.focus", {"bundle_id": bundle_id}, result_type=type(None))

    async def hide(self, bundle_id: str) -> None:
        """Hide a running application."""
        await self._api.call("apps.hide", {"bundle_id": bundle_id}, result_type=type(None))

    async def unhide(self, bundle_id: str) -> None:
        """Unhide a running application."""
        await self._api.call("apps.unhide", {"bundle_id": bundle_id}, result_type=type(None))

    async def quit(self, bundle_id: str) -> None:
        """Quit an application unless controls were disabled by configuration."""
        await self._api.call("apps.quit", {"bundle_id": bundle_id}, result_type=type(None))

File: hammerspoon/macapi/apps.lua
--- Application observation and control RPC service.
---
--- Methods are exported through the `methods` table and merged into the
--- dispatcher. Application quitting is guarded by the dangerous-action flag;
--- read-only inspection and focus/hide operations remain available.
local application = require("hs.application")
local validators = require("macapi.validators")
local common = require("macapi.common")
local config = require("macapi.config")
local M = { methods = {} }

local function find(bundle_id)
    local app = application.get(bundle_id)
    if not app then return common.error("APP_NOT_FOUND", "application was not found: " .. bundle_id) end
    return app
end

--- Return all currently running applications.
M.methods["apps.list"] = function()
    local result = {}
    for _, app in ipairs(application.runningApplications()) do
        table.insert(result, common.application_info(app))
    end
    return result
end

--- Return the frontmost application, if Hammerspoon reports one.
M.methods["apps.frontmost"] = function()
    local app = application.frontmostApplication()
    return app and common.application_info(app) or nil
end

--- Launch or focus an application identified by its bundle ID.
M.methods["apps.open"] = function(params)
    local bundle_id, error = validators.required_string(params, "bundle_id")
    if not bundle_id then return nil, error end
    local app = application.launchOrFocusByBundleID(bundle_id) and application.get(bundle_id)
    if not app then return common.error("APP_NOT_FOUND", "application could not be opened: " .. bundle_id) end
    return common.application_info(app)
end

local function action(params, name, callback)
    local bundle_id, error = validators.required_string(params, "bundle_id")
    if not bundle_id then return nil, error end
    local app, find_error = find(bundle_id)
    if not app then return nil, find_error end
    callback(app)
    return nil
end

--- Activate a running application.
M.methods["apps.focus"] = function(params)
    return action(params, "focus", function(app) app:activate(true) end)
end

--- Hide a running application.
M.methods["apps.hide"] = function(params)
    return action(params, "hide", function(app) app:hide() end)
end

--- Unhide a running application.
M.methods["apps.unhide"] = function(params)
    return action(params, "unhide", function(app) app:unhide() end)
end

--- Quit an application when dangerous actions are enabled.
M.methods["apps.quit"] = function(params)
    if not config.dangerous_actions_enabled then
        return common.error("FEATURE_DISABLED", "apps.quit is disabled by configuration")
    end
    return action(params, "quit", function(app) app:kill() end)
end

return M
